--- a/src/uinput.c
+++ b/src/uinput.c
@@ -112,6 +112,8 @@ static int dragskip = 0;
 static int touch_always = 0;
 static int touch_pressure = 1;
 static int abs_x = 0, abs_y = 0;
+static int multitouch = 0;
+static int tracking_id = 1;
 
 static char *devs[] = {
 	"/dev/misc/uinput",
@@ -703,6 +705,8 @@ void parse_uinput_str(char *in) {
 			}
 		} else if (strstr(p, "tslib_cal=") == p) {
 			tslib_cal = strdup(p+strlen("tslib_cal="));
+		} else if (strstr(p, "multitouch") == p) {
+			multitouch = 1;
 		} else {
 			rfbLog("invalid UINPUT option: %s\n", p);
 			clean_up_exit(1);
@@ -795,6 +799,18 @@ static void ptr_abs(int x, int y, int p) {
 	ev.value = x;
 	write(d, &ev, sizeof(ev));
 
+	if (multitouch) {
+		ev.type = EV_ABS;
+		ev.code = ABS_MT_POSITION_Y;
+		ev.value = y;
+		write(d, &ev, sizeof(ev));
+
+		ev.type = EV_ABS;
+		ev.code = ABS_MT_POSITION_X;
+		ev.value = x;
+		write(d, &ev, sizeof(ev));
+	}
+
 	if (p >= 0) {
 		ev.type = EV_ABS;
 		ev.code = ABS_PRESSURE;
@@ -807,6 +823,17 @@ static void ptr_abs(int x, int y, int p) {
 			ev.value = p ? 1 : 0;
 			write(d, &ev, sizeof(ev));
 		}
+
+		if (multitouch) {
+			ev.type = EV_ABS;
+			ev.code = ABS_MT_TRACKING_ID;
+			if (p == 0) {
+				ev.value = -1;
+			} else {
+				ev.value = tracking_id++;
+			}
+			write(d, &ev, sizeof(ev));
+		}
 	}
 
 	ev.type = EV_SYN;
