--- a/drivers/input/touchscreen/goodix.c
+++ b/drivers/input/touchscreen/goodix.c
@@ -1175,6 +1175,10 @@ retry_read_config:
 	/* Try overriding touchscreen parameters via device properties */
 	touchscreen_parse_properties(ts->input_dev, true, &ts->prop);
 
+	/* Some GT911 panels report valid resolution but 0 contact points. */
+	if (!ts->max_touch_num)
+		ts->max_touch_num = GOODIX_MAX_CONTACTS;
+
 	if (!ts->prop.max_x || !ts->prop.max_y || !ts->max_touch_num) {
 		if (!ts->reset_controller_at_probe &&
 		    ts->irq_pin_access_method != IRQ_PIN_ACCESS_NONE) {
@@ -1189,8 +1193,10 @@ retry_read_config:
 		dev_err(&ts->client->dev,
 			"Invalid config (%d, %d, %d), using defaults\n",
 			ts->prop.max_x, ts->prop.max_y, ts->max_touch_num);
-		ts->prop.max_x = GOODIX_MAX_WIDTH - 1;
-		ts->prop.max_y = GOODIX_MAX_HEIGHT - 1;
+		if (!ts->prop.max_x || !ts->prop.max_y) {
+			ts->prop.max_x = GOODIX_MAX_WIDTH - 1;
+			ts->prop.max_y = GOODIX_MAX_HEIGHT - 1;
+		}
 		ts->max_touch_num = GOODIX_MAX_CONTACTS;
 		input_abs_set_max(ts->input_dev,
 				  ABS_MT_POSITION_X, ts->prop.max_x);
